File: 1175 - Prime Arrangements.js
/**
 * @param {number} n
 * @return {number}
 */
let MOD = 10**9 + 7;
var numPrimeArrangements = function (n) {
    if (n == 1) return 1;
    let pCount = 0;
    for (let i = 2; i <= n; i++) {
        if (isPrime(i)) pCount++;
    }
    return multi(fac(pCount), fac(n - pCount));
};
function multi(a,b){
    //将b拆成2部分
    let t=Math.floor(b/100000),
        t2=b % 100000
    let sum=0
    for(let i=0;i<t;i++){
      sum=(sum+100000*a) % MOD
    }
    sum=(sum+t2*a)%MOD
    return sum
}

function isPrime(n) {
    for (let i = 2; i * i <= n; i++) {
        if (n % i == 0) return false;
    }
    return true;
}

function fac(n) {
    let res = 1;
    for (let i = 2; i <= n; i++) {
        res *= i;
        res %= MOD;
    }
    return res;
}

File: 1175 - Prime Arrangements.java
import java.io.Console;

class Solution {
    public int numPrimeArrangements(int n) {
        int primeCount = 0, m = 1000000000 + 7;
        if (n == 1)
            return 1;
        for (int i = 2; i <= n; i++) {
            if (isPrime(i))
                primeCount++;
        }

        long res = (fac(primeCount) * fac(n - primeCount));
        return (int)(res % m);

    }

    private long fac(int n) {
        long res = 1;
        for (int i = 1; i <= n; i++) {
            res *= i;
            res %= 1000000007;
        }
        return res;
    }

    private boolean isPrime(int n) {
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0)
                return false;
        }
        return true;
    }
}